File: src/main/java/dev/jlo/gamemodes/paper/PaperMatchService.java
package dev.jlo.gamemodes.paper;

import dev.jlo.gamemodes.domain.common.MatchPhase;
import dev.jlo.gamemodes.player.EffectSnapshot;
import dev.jlo.gamemodes.player.ItemStackSnapshot;
import dev.jlo.gamemodes.player.LocationSnapshot;
import dev.jlo.gamemodes.player.PendingRestore;
import dev.jlo.gamemodes.player.PendingRestoreRepository;
import dev.jlo.gamemodes.player.PlayerSnapshot;
import java.util.ArrayList;
import java.util.Base64;
import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.function.BiConsumer;
import java.util.function.BiFunction;
import java.util.function.Consumer;
import java.util.function.Function;
import org.bukkit.GameMode;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.NamespacedKey;
import org.bukkit.attribute.Attribute;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.potion.PotionEffect;
import org.bukkit.potion.PotionEffectType;

public class PaperMatchService implements CommandService {
    public static final class ArenaConfig {
        private final String id;
        private final Mode mode;
        private final int capacityPerTeam;
        private final int quorumPerTeam;

        public ArenaConfig(String id, Mode mode, int capacityPerTeam, int quorumPerTeam) {
            this.id = id;
            this.mode = mode;
            this.capacityPerTeam = capacityPerTeam;
            this.quorumPerTeam = quorumPerTeam;
        }

        public String getId() { return id; }
        public Mode getMode() { return mode; }
        public int getCapacityPerTeam() { return capacityPerTeam; }
        public int getQuorumPerTeam() { return quorumPerTeam; }
    }

    private final BiFunction<Mode, String, String> arenaProvider;
    private final Consumer<MatchSession> onSessionCreated;
    private final Consumer<MatchSession> onSessionRemoved;
    private final PendingRestoreRepository restoreRepository;
    private final Map<String, ArenaConfig> configuredArenas = new LinkedHashMap<>();
    private final Map<String, MatchSession> sessions = new LinkedHashMap<>();
    private final Map<UUID, MatchSession> owners = new LinkedHashMap<>();
    private final Map<UUID, PlayerSnapshot> snapshots = new LinkedHashMap<>();
    private final SafetyPolicy safety = new SafetyPolicy();
    private Consumer<UUID> captureHook = ignored -> {};
    private Consumer<UUID> restoreHook = ignored -> {};
    private TriConsumer<UUID, MatchSession, TeamAssignment> admitHook = (a, b, c) -> {};
    private Function<Void, String> reloadHook = ignored -> "No arena catalog is installed";

    public PaperMatchService() {
        this((mode, arena) -> arena, ignored -> {}, ignored -> {}, null);
    }

    public PaperMatchService(BiFunction<Mode, String, String> arenaProvider,
                             Consumer<MatchSession> onSessionCreated,
                             Consumer<MatchSession> onSessionRemoved,
                             PendingRestoreRepository restoreRepository) {
        this.arenaProvider = arenaProvider;
        this.onSessionCreated = onSessionCreated;
        this.onSessionRemoved = onSessionRemoved;
        this.restoreRepository = restoreRepository;
    }

    @FunctionalInterface
    public interface TriConsumer<A, B, C> { void accept(A a, B b, C c); }

    public static final class ArenaConfigBuilder {}

    public synchronized void configureArenas(Collection<ArenaConfig> arenas) {
        configuredArenas.clear();
        for (ArenaConfig arena : arenas) {
            if (!(arena.getCapacityPerTeam() > 0 && arena.getQuorumPerTeam() >= 1
                    && arena.getQuorumPerTeam() <= arena.getCapacityPerTeam())) {
                throw new IllegalArgumentException("Failed requirement.");
            }
            if (configuredArenas.putIfAbsent(arena.getId(), arena) != null) {
                throw new IllegalArgumentException("Duplicate arena ID " + arena.getId());
            }
        }
    }

    public void installPlayerHooks(Consumer<UUID> capture, Consumer<UUID> restore,
                                   TriConsumer<UUID, MatchSession, TeamAssignment> admit) {
        captureHook = capture;
        restoreHook = restore;
        admitHook = admit;
    }

    public void installReloadHook(SupplierString reload) { reloadHook = ignored -> reload.get(); }

    @FunctionalInterface
    public interface SupplierString { String get(); }

    public SafetyPolicy safetyPolicy() { return safety; }
    public List<MatchSession> sessions() { return new ArrayList<>(sessions.values()); }
    public ArenaConfig arenaConfig(String id) { return configuredArenas.get(id); }

    @Override
    public TeamAssignment join(UUID player, Mode mode, String arena) {
        if (owners.containsKey(player)) throw new IllegalStateException("Player is already queued or in a match");
        String arenaId = arenaProvider.apply(mode, arena);
        if (arenaId == null) throw new IllegalStateException("No configured " + mode.name() + " arena is available");
        ArenaConfig configured = configuredArenas.get(arenaId);
        if (configured == null) throw new IllegalStateException("Arena '" + arenaId + "' is not configured");
        if (configured.getMode() != mode) throw new IllegalArgumentException("Arena '" + arenaId + "' is unavailable for " + mode.name());
        MatchSession session = sessions.values().stream()
                .filter(it -> it.getMode() == mode && it.getArenaId().equals(arenaId) && it.getPhase() == MatchPhase.WAITING)
                .findFirst().orElseGet(() -> {
                    MatchSession created = new MatchSession(mode, arenaId, configured.getCapacityPerTeam(), configured.getQuorumPerTeam());
                    onSessionCreated.accept(created);
                    sessions.put(created.getId(), created);
                    return created;
                });
        captureHook.accept(player);
        try {
            TeamAssignment team = session.join(player);
            owners.put(player, session);
            admitHook.accept(player, session, team);
            safety.activate(player);
            return team;
        } catch (RuntimeException failure) {
            owners.remove(player);
            safety.deactivate(player);
            if (session.teamOf(player) != null) session.leave(player);
            restoreHook.accept(player);
            throw failure;
        }
    }

    @Override
    public boolean leave(UUID player) {
        MatchSession session = owners.remove(player);
        if (session == null) return false;
        session.leave(player);
        safety.deactivate(player);
        restoreHook.accept(player);
        return true;
    }

    @Override public boolean ready(UUID player) { MatchSession s = owners.get(player); return s != null && s.ready(player); }
    @Override public String status(UUID player) { MatchSession s = owners.get(player); return s == null ? "NONE" : s.getMode().name() + " " + s.getPhase().name(); }
    @Override public TeamAssignment team(UUID player) { MatchSession s = owners.get(player); return s == null ? null : s.teamOf(player); }

    @Override
    public String admin(UUID player, String action, List<String> args) {
        return switch (action.toLowerCase()) {
            case "start" -> {
                String arena = args.isEmpty() ? null : args.get(0);
                boolean started = sessions.values().stream().filter(s -> arena == null || s.getArenaId().equals(arena)).anyMatch(MatchSession::startIfReady);
                yield started ? "started" : "not ready";
            }
            case "stop" -> {
                String arena = args.isEmpty() ? null : args.get(0);
                List<MatchSession> selected = sessions.values().stream().filter(s -> arena == null || s.getArenaId().equals(arena)).toList();
                selected.forEach(this::terminate);
                yield "stopped " + selected.size();
            }
            case "arena" -> {
                String sub = args.isEmpty() ? null : args.get(0).toLowerCase();
                if (sub == null || sub.equals("list")) {
                    StringBuilder result = new StringBuilder("arenas: ");
                    boolean first = true;
                    for (ArenaConfig a : configuredArenas.values()) { if (!first) result.append(", "); first = false; result.append(a.getId()).append('(').append(a.getMode()).append(')'); }
                    yield result.toString();
                }
                yield sub.equals("validate") ? "validated " + configuredArenas.size() + " arenas" : "Usage: arena <list|validate>";
            }
            case "reload" -> reloadHook.apply(null);
            case "debug" -> {
                StringBuilder result = new StringBuilder("sessions: ");
                boolean first = true;
                for (MatchSession s : sessions.values()) { if (!first) result.append(", "); first = false; result.append(s.getMode()).append('/').append(s.getArenaId()).append('/').append(s.getPhase()); }
                yield result.toString();
            }
            default -> "unknown admin action";
        };
    }

    private void terminate(MatchSession session) {
        List<UUID> players = owners.entrySet().stream().filter(e -> e.getValue() == session).map(Map.Entry::getKey).toList();
        session.cleanup();
        for (UUID player : players) { safety.deactivate(player); restoreHook.accept(player); owners.remove(player); }
        sessions.remove(session.getId());
        onSessionRemoved.accept(session);
    }

    public void tick() {
        for (MatchSession session : new ArrayList<>(sessions.values())) {
            session.startIfReady(); session.advance();
            if (session.getPhase() == MatchPhase.RESOLVING || session.getPhase() == MatchPhase.CLEANUP) terminate(session);
        }
    }

    public void shutdown(Function<UUID, Player> players) {
        for (MatchSession session : new ArrayList<>(sessions.values())) terminate(session);
        for (UUID player : new ArrayList<>(owners.keySet())) {
            Player p = players.apply(player);
            if (p != null) restore(p); else restoreHook.accept(player);
            safety.deactivate(player);
        }
        owners.clear(); sessions.clear();
    }

    public boolean isParticipant(UUID player) { return owners.containsKey(player); }
    public void disconnect(UUID player) { MatchSession s = owners.get(player); if (s != null) s.disconnect(player); }
    public boolean recordCombat(UUID attacker, UUID victim) { MatchSession s = owners.get(attacker); return s != null && owners.get(victim) == s && s.recordCombat(attacker, victim); }
    public boolean playerDied(UUID player, UUID killer) { MatchSession s = owners.get(player); if (s == null) return false; if (killer != null && owners.get(killer) != s) return s.playerDied(player); return s.playerDied(player, killer); }
    public Location respawn(Player player) { MatchSession s = owners.get(player.getUniqueId()); return s == null ? null : s.respawnLocation(player); }

    public void capture(Player player) {
        List<ItemStackSnapshot> inventory = new ArrayList<>(); for (ItemStack item : player.getInventory().getStorageContents()) inventory.add(itemSnapshot(item));
        ItemStack cursor = player.getItemOnCursor(); ItemStack armor = null;
        List<ItemStackSnapshot> armorSnapshots = new ArrayList<>(); for (ItemStack item : player.getInventory().getArmorContents()) armorSnapshots.add(itemSnapshot(item));
        ItemStack offhand = player.getInventory().getItemInOffHand();
        List<EffectSnapshot> effects = new ArrayList<>();
        for (PotionEffect effect : player.getActivePotionEffects()) effects.add(new EffectSnapshot(effect.getType().getKey().toString(), effect.getAmplifier(), effect.getDuration(), effect.isAmbient(), effect.hasParticles(), effect.hasIcon()));
        double maxHealth = player.getAttribute(Attribute.MAX_HEALTH) == null ? player.getHealth() : player.getAttribute(Attribute.MAX_HEALTH).getBaseValue();
        Location location = player.getLocation();
        PlayerSnapshot snapshot = new PlayerSnapshot(player.getUniqueId(), inventory, itemSnapshotOrNull(cursor), armorSnapshots, itemSnapshotOrNull(offhand), effects, Map.of("minecraft:max_health", maxHealth), player.getHealth(), player.getFoodLevel(), player.getSaturation(), player.getExp(), player.getLevel(), player.getGameMode().name(), player.getAllowFlight(), player.isFlying(), new LocationSnapshot(location.getWorld().getName(), location.getX(), location.getY(), location.getZ(), location.getYaw(), location.getPitch()));
        if (restoreRepository != null) restoreRepository.put(snapshot); else snapshots.put(player.getUniqueId(), snapshot);
        player.getInventory().clear(); player.getInventory().setArmorContents(new ItemStack[0]); player.setItemOnCursor(new ItemStack(Material.AIR));
        for (PotionEffect effect : new ArrayList<>(player.getActivePotionEffects())) player.removePotionEffect(effect.getType());
        player.setGameMode(GameMode.ADVENTURE); player.setAllowFlight(false); player.setFlying(false);
    }

    private ItemStackSnapshot itemSnapshotOrNull(ItemStack item) { return item == null || item.getType().isAir() ? null : itemSnapshot(item); }
    private ItemStackSnapshot itemSnapshot(ItemStack item) { if (item == null || item.getType().isAir()) return new ItemStackSnapshot(Material.AIR.name(), 0); return new ItemStackSnapshot(item.getType().name(), item.getAmount(), Base64.getEncoder().encodeToString(item.serializeAsBytes())); }

    public void restore(Player player) {
        PendingRestore pending = restoreRepository == null ? null : restoreRepository.claim(player.getUniqueId());
        PlayerSnapshot snapshot = pending == null ? snapshots.remove(player.getUniqueId()) : pending.getSnapshot(); if (snapshot == null) return;
        try {
            player.getInventory().setStorageContents(snapshot.getInventory().stream().map(this::restoreItem).toArray(ItemStack[]::new));
            player.getInventory().setArmorContents(snapshot.getArmor().stream().map(this::restoreItem).toArray(ItemStack[]::new));
            player.getInventory().setItemInOffHand(restoreItem(snapshot.getOffhand()) == null ? new ItemStack(Material.AIR) : restoreItem(snapshot.getOffhand())); player.setItemOnCursor(restoreItem(snapshot.getCursor()) == null ? new ItemStack(Material.AIR) : restoreItem(snapshot.getCursor()));
            for (PotionEffect effect : new ArrayList<>(player.getActivePotionEffects())) player.removePotionEffect(effect.getType());
            for (EffectSnapshot effect : snapshot.getEffects()) { NamespacedKey key = NamespacedKey.fromString(effect.getType()); if (key == null) continue; PotionEffectType type = PotionEffectType.getByKey(key); if (type != null) player.addPotionEffect(new PotionEffect(type, effect.getDurationTicks(), effect.getAmplifier(), effect.getAmbient(), effect.getParticles(), effect.getIcon())); }
            Double max = snapshot.getAttributes().get("minecraft:max_health"); if (max != null && player.getAttribute(Attribute.MAX_HEALTH) != null) player.getAttribute(Attribute.MAX_HEALTH).setBaseValue(max);
            double currentMax = player.getAttribute(Attribute.MAX_HEALTH) == null ? snapshot.getHealth() : player.getAttribute(Attribute.MAX_HEALTH).getValue(); player.setHealth(Math.min(snapshot.getHealth(), currentMax)); player.setFoodLevel(snapshot.getFood()); player.setSaturation((float) snapshot.getSaturation()); player.setExp(snapshot.getExperience()); player.setLevel(snapshot.getLevel()); player.setGameMode(GameMode.valueOf(snapshot.getGameMode())); player.setAllowFlight(snapshot.getAllowFlight()); player.setFlying(snapshot.getFlying());
            LocationSnapshot location = snapshot.getReturnLocation(); if (location != null) { var server = player.getServer(); var world = server.getWorld(location.getWorld()); if (world == null) world = server.getWorlds().get(0); player.teleport(new Location(world, location.getX(), location.getY(), location.getZ(), location.getYaw(), location.getPitch())); }
            if (pending != null && !restoreRepository.markRestored(player.getUniqueId())) throw new IllegalStateException("Failed to mark player restore complete");
        } catch (RuntimeException failure) { if (pending != null) restoreRepository.markFailed(player.getUniqueId(), failure.getMessage() == null ? failure.getClass().getSimpleName() : failure.getMessage()); throw failure; }
    }

    private ItemStack restoreItem(ItemStackSnapshot snapshot) { if (snapshot == null || snapshot.getAmount() <= 0 || snapshot.getMaterial().equals(Material.AIR.name())) return null; return snapshot.getMetadata() != null && !snapshot.getMetadata().isEmpty() ? ItemStack.deserializeBytes(Base64.getDecoder().decode(snapshot.getMetadata())) : new ItemStack(Material.valueOf(snapshot.getMaterial()), snapshot.getAmount()); }
}

File: src/main/kotlin/dev/jlo/gamemodes/paper/PaperMatchService.kt
package dev.jlo.gamemodes.paper

import dev.jlo.gamemodes.domain.common.MatchPhase
import dev.jlo.gamemodes.player.EffectSnapshot
import dev.jlo.gamemodes.player.ItemStackSnapshot
import dev.jlo.gamemodes.player.LocationSnapshot
import dev.jlo.gamemodes.player.PendingRestoreRepository
import dev.jlo.gamemodes.player.PlayerSnapshot
import java.util.Base64
import java.util.UUID
import org.bukkit.GameMode
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.NamespacedKey
import org.bukkit.attribute.Attribute
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import org.bukkit.potion.PotionEffect
import org.bukkit.potion.PotionEffectType

class PaperMatchService(
    private val arenaProvider: (Mode, String?) -> String? = { _, arena -> arena },
    private val onSessionCreated: (MatchSession) -> Unit = {},
    private val onSessionRemoved: (MatchSession) -> Unit = {},
    private val restoreRepository: PendingRestoreRepository? = null
) : CommandService {
    data class ArenaConfig(val id: String, val mode: Mode, val capacityPerTeam: Int, val quorumPerTeam: Int)
    private val configuredArenas = linkedMapOf<String, ArenaConfig>()
    private val sessions = linkedMapOf<String, MatchSession>()
    private val owners = linkedMapOf<UUID, MatchSession>()
    private val snapshots = linkedMapOf<UUID, PlayerSnapshot>()
    private val safety = SafetyPolicy()
    private var captureHook: (UUID) -> Unit = {}
    private var restoreHook: (UUID) -> Unit = {}
    private var admitHook: (UUID, MatchSession, TeamAssignment) -> Unit = { _, _, _ -> }
    private var reloadHook: () -> String = { "No arena catalog is installed" }

    @Synchronized
    fun configureArenas(arenas: Collection<ArenaConfig>) {
        configuredArenas.clear()
        arenas.forEach {
            require(it.capacityPerTeam > 0 && it.quorumPerTeam in 1..it.capacityPerTeam)
            require(configuredArenas.putIfAbsent(it.id, it) == null) { "Duplicate arena ID ${it.id}" }
        }
    }

    fun installPlayerHooks(
        capture: (UUID) -> Unit,
        restore: (UUID) -> Unit,
        admit: (UUID, MatchSession, TeamAssignment) -> Unit
    ) {
        captureHook = capture
        restoreHook = restore
        admitHook = admit
    }
    fun installReloadHook(reload: () -> String) {
        reloadHook = reload
    }
    fun safetyPolicy(): SafetyPolicy = safety
    fun sessions(): List<MatchSession> = sessions.values.toList()
    fun arenaConfig(id: String): ArenaConfig? = configuredArenas[id]

    override fun join(player: UUID, mode: Mode, arena: String?): TeamAssignment {
        check(player !in owners) { "Player is already queued or in a match" }
        val arenaId = arenaProvider(mode, arena) ?: error("No configured ${mode.name} arena is available")
        val configured = configuredArenas[arenaId] ?: error("Arena '$arenaId' is not configured")
        require(configured.mode == mode) { "Arena '$arenaId' is unavailable for ${mode.name}" }
        val session = sessions.values.firstOrNull {
            it.mode == mode && it.arenaId == arenaId && it.phase == MatchPhase.WAITING
        } ?: MatchSession(mode, arenaId, configured.capacityPerTeam, configured.quorumPerTeam).also {
            onSessionCreated(it)
            sessions[it.id] = it
        }
        captureHook(player)
        try {
            val team = session.join(player)
            owners[player] = session
            admitHook(player, session, team)
            safety.activate(player)
            return team
        } catch (failure: RuntimeException) {
            owners.remove(player)
            safety.deactivate(player)
            if (session.teamOf(player) != null) session.leave(player)
            restoreHook(player)
            throw failure
        }
    }

    override fun leave(player: UUID): Boolean {
        val session = owners.remove(player) ?: return false
        session.leave(player)
        safety.deactivate(player)
        restoreHook(player)
        return true
    }

    override fun ready(player: UUID): Boolean = owners[player]?.ready(player) ?: false

    override fun status(player: UUID): String = owners[player]?.let { "${it.mode.name} ${it.phase.name}" } ?: "NONE"

    override fun team(player: UUID): TeamAssignment? = owners[player]?.teamOf(player)

    override fun admin(player: UUID?, action: String, args: List<String>): String = when (action.lowercase()) {
        "start" -> {
            val arena = args.firstOrNull()
            val started = sessions.filterValues { arena == null || it.arenaId == arena }.values.any { it.startIfReady() }
            if (started) "started" else "not ready"
        }
        "stop" -> {
            val selected = sessions.filterValues { args.firstOrNull() == null || it.arenaId == args.first() }.values.toList()
            selected.forEach(::terminate)
            "stopped ${selected.size}"
        }
        "arena" -> when (args.firstOrNull()?.lowercase()) {
            null, "list" -> configuredArenas.values.joinToString(prefix = "arenas: ") { "${it.id}(${it.mode})" }
            "validate" -> "validated ${configuredArenas.size} arenas"
            else -> "Usage: arena <list|validate>"
        }
        "reload" -> reloadHook()
        "debug" -> sessions.values.joinToString(prefix = "sessions: ") { "${it.mode}/${it.arenaId}/${it.phase}" }
        else -> "unknown admin action"
    }

    private fun terminate(session: MatchSession) {
        val players = owners.filterValues { it === session }.keys.toList()
        session.cleanup()
        players.forEach { player ->
            safety.deactivate(player)
            restoreHook(player)
            owners.remove(player)
        }
        sessions.remove(session.id)
        onSessionRemoved(session)
    }

    fun tick() {
        sessions.values.toList().forEach { session ->
            session.startIfReady()
            session.advance()
            if (session.phase == MatchPhase.RESOLVING || session.phase == MatchPhase.CLEANUP) terminate(session)
        }
    }

    fun shutdown(players: (UUID) -> Player?) {
        sessions.values.toList().forEach(::terminate)
        owners.keys.toList().forEach { player ->
            players(player)?.let { restore(it) } ?: restoreHook(player)
            safety.deactivate(player)
        }
        owners.clear()
        sessions.clear()
    }
    fun isParticipant(player: UUID): Boolean = owners.containsKey(player)

    fun disconnect(player: UUID) {
        owners[player]?.disconnect(player)
    }

    fun recordCombat(attacker: UUID, victim: UUID): Boolean {
        val session = owners[attacker] ?: return false
        if (owners[victim] !== session) return false
        return session.recordCombat(attacker, victim)
    }

    fun playerDied(player: UUID, killer: UUID? = null): Boolean {
        val session = owners[player] ?: return false
        if (killer != null && owners[killer] !== session) return session.playerDied(player)
        return session.playerDied(player, killer)
    }

    fun respawn(player: Player): Location? = owners[player.uniqueId]?.respawnLocation(player)

    fun capture(player: Player) {
        val snapshot = PlayerSnapshot(
            playerId = player.uniqueId,
            inventory = player.inventory.storageContents.map(::itemSnapshot),
            cursor = player.itemOnCursor.takeUnless { it.type.isAir }?.let(::itemSnapshot),
            armor = player.inventory.armorContents.map(::itemSnapshot),
            offhand = player.inventory.itemInOffHand.takeUnless { it.type.isAir }?.let(::itemSnapshot),
            effects = player.activePotionEffects.map {
                EffectSnapshot(
                    it.type.key.toString(), it.amplifier, it.duration,
                    it.isAmbient, it.hasParticles(), it.hasIcon()
                )
            },
            attributes = mapOf(
                "minecraft:max_health" to
                    (player.getAttribute(Attribute.MAX_HEALTH)?.baseValue ?: player.health)
            ),
            health = player.health,
            food = player.foodLevel,
            saturation = player.saturation.toDouble(),
            experience = player.exp,
            level = player.level,
            gameMode = player.gameMode.name,
            allowFlight = player.allowFlight,
            flying = player.isFlying,
            returnLocation = player.location.let {
                LocationSnapshot(it.world.name, it.x, it.y, it.z, it.yaw, it.pitch)
            }
        )
        if (restoreRepository != null) restoreRepository.put(snapshot) else snapshots[player.uniqueId] = snapshot
        player.inventory.clear()
        player.inventory.armorContents = emptyArray()
        player.setItemOnCursor(ItemStack(Material.AIR))
        player.activePotionEffects.forEach { player.removePotionEffect(it.type) }
        player.gameMode = GameMode.ADVENTURE
        player.allowFlight = false
        player.isFlying = false
    }

    fun restore(player: Player) {
        val pending = restoreRepository?.claim(player.uniqueId)
        val snapshot = pending?.snapshot ?: snapshots.remove(player.uniqueId) ?: return
        try {
            player.inventory.storageContents = snapshot.inventory.map(::restoreItem).toTypedArray()
            player.inventory.armorContents = snapshot.armor.map(::restoreItem).toTypedArray()
            player.inventory.setItemInOffHand(restoreItem(snapshot.offhand) ?: ItemStack(Material.AIR))
            player.setItemOnCursor(restoreItem(snapshot.cursor) ?: ItemStack(Material.AIR))
            player.activePotionEffects.forEach { player.removePotionEffect(it.type) }
            snapshot.effects.forEach { effect ->
                val key = NamespacedKey.fromString(effect.type) ?: return@forEach
                val type = PotionEffectType.getByKey(key) ?: return@forEach
                player.addPotionEffect(
                    PotionEffect(
                        type, effect.durationTicks, effect.amplifier,
                        effect.ambient, effect.particles, effect.icon
                    )
                )
            }
            snapshot.attributes["minecraft:max_health"]?.let { value ->
                player.getAttribute(Attribute.MAX_HEALTH)?.baseValue = value
            }
            player.health = snapshot.health.coerceAtMost(player.getAttribute(Attribute.MAX_HEALTH)?.value ?: snapshot.health)
            player.foodLevel = snapshot.food
            player.saturation = snapshot.saturation.toFloat()
            player.exp = snapshot.experience
            player.level = snapshot.level
            player.gameMode = GameMode.valueOf(snapshot.gameMode)
            player.allowFlight = snapshot.allowFlight
            player.isFlying = snapshot.flying
            snapshot.returnLocation?.let { location ->
                val world = player.server.getWorld(location.world) ?: player.server.worlds.first()
                player.teleport(Location(world, location.x, location.y, location.z, location.yaw, location.pitch))
            }
            if (pending != null) check(restoreRepository!!.markRestored(player.uniqueId))
        } catch (failure: RuntimeException) {
            if (pending != null) restoreRepository!!.markFailed(player.uniqueId, failure.message ?: failure.javaClass.simpleName)
            throw failure
        }
    }

    private fun itemSnapshot(item: ItemStack?): ItemStackSnapshot {
        if (item == null || item.type.isAir) return ItemStackSnapshot(Material.AIR.name, 0)
        return ItemStackSnapshot(
            item.type.name,
            item.amount,
            Base64.getEncoder().encodeToString(item.serializeAsBytes())
        )
    }

    private fun restoreItem(snapshot: ItemStackSnapshot?): ItemStack? {
        if (snapshot == null || snapshot.amount <= 0 || snapshot.material == Material.AIR.name) return null
        return if (snapshot.metadata.isNotEmpty()) {
            ItemStack.deserializeBytes(Base64.getDecoder().decode(snapshot.metadata))
        } else {
            ItemStack(Material.valueOf(snapshot.material), snapshot.amount)
        }
    }
}
